### PR TITLE
feat: AmazonBedrockChatGenerator - add tools support

### PR DESCRIPTION
**Pull request by vblagoje**
### Why:
Adds Haystack tool support to `AmazonBedrockChatGenerator`

- fixes https://github.com/deepset-ai/haystack/issues/8732

### What:
- Added method to convert Haystack tools to Amazon Bedrock toolConfig format, facilitating tool integration.
- Modified `AmazonBedrockChatGenerator` to accept, serialize, and handle tool specifications as part of the chat messages.
- Updated testing suite to cover tool-related functionalities, ensuring robustness and reliability.

### How can it be used:
The changes enable users to pass tools to `AmazonBedrockChatGenerator`, enhancing the model's capabilities by allowing it to use external tools within chat interactions. Here is a simplified example:
```python
from haystack.dataclasses import ChatMessage
from haystack_integrations.components.generators.amazon_bedrock import AmazonBedrockChatGenerator
from haystack.tools import Tool

def weather(city: str):
        """Get weather for a given city."""
        return f"The weather in {city} is sunny and 32°C"

tool_parameters = {"type": "object", "properties": {"city": {"type": "string"}}, "required": ["city"]}
tool = Tool(
    name="weather",
    description="useful to determine the weather in a given location",
    parameters=tool_parameters,
    function=weather,
)

messages = [ChatMessage.from_user("What's the weather like in Paris?")]
generator = AmazonBedrockChatGenerator(model="model_name", tools=[tool])
response = generator.run(messages=messages)
```

### How did you test it:
Implemented thorough unit and integration tests to validate the addition of tool-related functionalities. Tests included:
- Serialization and deserialization of tool configurations.
- Ensuring tool calls within conversations behave as expected.
- Integration tests with mock and live scenarios for tool usage within chat interactions.

### Notes for the reviewer:
Please pay special attention to the serialization and deserialization processes for tool objects and the handling of tool calls within the chat messages. Consider reviewing the edge cases in the test suite that handle different orderings of tool messages in replies.

### CODE REVIEW

File: integrations/amazon_bedrock/src/haystack_integrations/components/generators/amazon_bedrock/chat/chat_generator.py [hunk 1]
```diff
@@ -19,6 +20,215 @@
...
+                except json.JSONDecodeError:
+                    # Keep as string if not valid JSON
+                    pass
+
+                tool_call = ToolCall(
+                    id=current_tool_call["id"],
+                    tool_name=current_tool_call["name"],
+                    arguments=current_tool_call["arguments"],
+                )
+                replies.append(ChatMessage.from_assistant("", tool_calls=[tool_call], meta=base_meta.copy()))
```
Comment (anakin87) [inline integrations/amazon_bedrock/src/haystack_integrations/components/generators/amazon_bedrock/chat/chat_generator.py:208]: same comment as above: `ChatMessage.from_assistant(text="", ...)` is handled properly by Bedrock?

File: integrations/amazon_bedrock/tests/test_chat_generator.py [hunk 2]
```diff
@@ -447,12 +454,170 @@ def test_callback(chunk: StreamingChunk):
...
+            data={"generator": {"messages": [ChatMessage.from_user("What's the weather like in Paris?")]}}
+        )
+
+        assert (
+            "The weather in Paris is sunny and 32°C"
+            == results["tool_invoker"]["tool_messages"][0].tool_call_result.result
+        )
+
+    @pytest.mark.parametrize("model_name", [MODELS_TO_TEST_WITH_TOOLS[0]])  # just one model is enough
+    @pytest.mark.integration
```
Comment (anakin87) [inline integrations/amazon_bedrock/tests/test_chat_generator.py:578]: I would transform this into a simpler unit test about serde (without running the pipeline).
Pipeline running with tools is already tested in `test_pipeline_with_amazon_bedrock_chat_generator`.

**Review by anakin87 — COMMENTED** [review]
I have some questions and small suggestions.

In general, this PR looks good.

File: integrations/amazon_bedrock/src/haystack_integrations/components/generators/amazon_bedrock/chat/chat_generator.py [hunk 3]
```diff
@@ -19,6 +20,215 @@
...
+                    tool_use = content_block["toolUse"]
+                    tool_call = ToolCall(
+                        id=tool_use.get("toolUseId"),
+                        tool_name=tool_use.get("name"),
+                        arguments=tool_use.get("input", {}),
+                    )
+                    tool_calls.append(tool_call)
+
+            # Create a single ChatMessage with combined text and tool calls
+            replies.append(ChatMessage.from_assistant(" ".join(text_content), tool_calls=tool_calls, meta=base_meta))
```
[the remark this answers is not in the document]
Comment (vblagoje) [reply integrations/amazon_bedrock/src/haystack_integrations/components/generators/amazon_bedrock/chat/chat_generator.py:142]: Yeah, not here, both "" and None work without an issue across all tested models. Let me know if I should change this? I'll keep it as is unless you advise otherwise

File: integrations/amazon_bedrock/src/haystack_integrations/components/generators/amazon_bedrock/chat/chat_generator.py [hunk 4]
```diff
@@ -19,6 +20,215 @@
...
+    :return: List of ChatMessage objects
+    """
+    replies = []
+    if "output" in response_body and "message" in response_body["output"]:
+        message = response_body["output"]["message"]
+        if message["role"] == "assistant":
+            content_blocks = message["content"]
+
+            # Common meta information
+            base_meta = {
```
[the remark this answers is not in the document]
Comment (vblagoje) [reply integrations/amazon_bedrock/src/haystack_integrations/components/generators/amazon_bedrock/chat/chat_generator.py:113]: No objection but metadata is not very rich:
```
{'ResponseMetadata': {'RequestId': 'c86b7b53-838e-4cef-b884-67739cc4f0db', 'HTTPStatusCode': 200, 'HTTPHeaders': {'date': 'Thu, 23 Jan 2025 14:47:19 GMT', 'content-type': 'application/json', 'content-length': '186', 'connection': 'keep-alive', 'x-amzn-requestid': 'c86b7b53-838e-4cef-b884-67739cc4f0db'}, 'RetryAttempts': 0}, 'output': {'message': {'role': 'assistant', 'content': [{'text': 'Paris.'}]}}, 'stopReason': 'end_turn', 'usage': {'inputTokens': 28, 'outputTokens': 3, 'totalTokens': 31}, 'metrics': {'latencyMs': 131}}
```

File: integrations/amazon_bedrock/src/haystack_integrations/components/generators/amazon_bedrock/chat/chat_generator.py (hunk 1 above)
Comment (vblagoje) [reply integrations/amazon_bedrock/src/haystack_integrations/components/generators/amazon_bedrock/chat/chat_generator.py:208]: Yes, double checked this - no issues, see above for details

**Review by vblagoje — COMMENTED** [review]
Ah missing the pydoc class tooling example. I'll add it @anakin87

**Comment by vblagoje** [comment]
@anakin87 lmk what we should do about ids or perhaps we can do that in separate PR? I'm ok with both

File: integrations/amazon_bedrock/tests/test_chat_generator.py (hunk 2 above)
Comment (anakin87) [reply integrations/amazon_bedrock/tests/test_chat_generator.py:578]: please simplify this test

**Review by anakin87 — COMMENTED** [review]
At this point, let's do nothing about ID and wait for @lambda-science's opinion in the issue. 

Please take this comment into account: https://github.com/deepset-ai/haystack-core-integrations/pull/1304#discussion_r1926707626
Then this PR is ready to be merged.

**Comment by vblagoje** [comment]
> At this point, let's do nothing about ID and wait for @lambda-science's opinion in the issue.
> 
> Please take this comment into account: [#1304 (comment)](https://github.com/deepset-ai/haystack-core-integrations/pull/1304#discussion_r1926707626) Then this PR is ready to be merged.

Right - I forgot about that one 🙏 
Will correct now

**Comment by vblagoje** [comment]
Ok should be gtg now @anakin87  See https://github.com/deepset-ai/haystack-core-integrations/pull/1304/commits/91f23304a5d3d40fe90e8d19f116f88d95e20c25

File: integrations/amazon_bedrock/tests/test_chat_generator.py [hunk 5]
```diff
@@ -447,12 +454,162 @@ def test_callback(chunk: StreamingChunk):
...
```
Comment (anakin87) [inline integrations/amazon_bedrock/tests/test_chat_generator.py:578]:
```suggestion
```

- let's transform this into a unit test (with mocking?)
- if possible, let's test the dictionary format more extensively
- let's also dump to YAML and reload from YAML (some serialization errors happen in this phase)

An example of what I mean: https://github.com/deepset-ai/haystack-core-integrations/blob/68ec202baa7cfa9af12bcf610498fbfeba5bc299/integrations/google_ai/tests/generators/chat/test_chat_gemini.py#L258

**Review by anakin87 — CHANGES_REQUESTED** [review]
last change request (hopefully 😊)

**Comment by vblagoje** [comment]
Ok, I agree,  how about now? 🤞
